src/svc_dg.c

```diff
@@ -330,7 +359,8 @@ svc_dg_recv(struct svc_req *req)
 	if (rlen == -1 || (rlen < (ssize_t) (4 * sizeof(u_int32_t))))
 		return (false);
 
-	xprt->xp_remote.nb.len = mesgp->msg_namelen;
+	/* copy the address */
+	__rpc_copy_address(&xprt->xp_remote, &ss, mesgp->msg_namelen);
```

Comment:
I'm not sure I understand this change.  It looks like it can only make a difference if msg_namelen is zero, but this should not happen, according to the man page.  Is this returning zero somewhere?

Reply:
That patch is NOT up for review! Anyway, we were using xprt->xp_remote to fill in the client's IP address. This is read by a worker thread and at the same time modified by svc_dg_recv() for receiving the next request.  svc_dg_recv() sets sa_family to 0xffff and this is read by the worker thread. Worker threads asserts when this happens. Happened when we have many clients doing NFSv3 mounts as they do UDP protocol for mounting. We were able to recreate this defect by using UDP for NFS (NFSv3) with a single client one  "dd" writing to a file.

This patch gets the remote address into a stack variable (ss) and it is copied into xprt later by calling __rpc_copy_address(). Hopefully, we copy the same data every time! I don't understand enough about UDP to know if this is the correct fix though.

Reply:
Okay.  The last patch (the change to the output threads) looks fine.